### DISCUSSION TITLE
Some way to configure GPIOs for private builds.

### DISCUSSION THREAD

**Discussion by mikep1998** [arendst/Tasmota#10198] [Q&A] [answered]
I would like a way to configure GPIOs into my private builds. 
Right now I use a my own driver and during the FUNC_SETTINGS_OVERRIDE event. I configure GPIOs for the device I'm building for. 
Like this:

    SettingsUpdateText(SET_TEMPLATE_NAME, "jacuzzi_Tub_Down");

    Settings.user_template.gp.io[0]  = AGPIO(GPIO_USER);         // 0       (I)O                 GPIO0, ADC2_CH1, TOUCH1, RTC_GPIO11, CLK_OUT1, EMAC_TX_CLK
    Settings.user_template.gp.io[1]  = AGPIO(GPIO_USER);         // 1       IO     TXD0          GPIO1, U0TXD, CLK_OUT3, EMAC_RXD2
    Settings.user_template.gp.io[2]  = AGPIO(GPIO_USER);         // 2    *  IO                   GPIO2, ADC2_CH2, TOUCH2, RTC_GPIO12, HSPIWP, HS2_DATA0, SD_DATA0
    Settings.user_template.gp.io[3]  = AGPIO(GPIO_USER);         // 3       IO     RXD0          GPIO3, U0RXD, CLK_OUT2
    Settings.user_template.gp.io[4]  = AGPIO(GPIO_SWT1_NP) +1;   // 4    *  IO   Water in Drain  GPIO4, ADC2_CH0, TOUCH0, RTC_GPIO10, HSPIHD, HS2_DATA1, SD_DATA1, EMAC_TX_ER
    Settings.user_template.gp.io[5]  = AGPIO(GPIO_SSPI_CS);      // 5    *  IO   SPI CS          GPIO5, VSPICS0, HS1_DATA6, EMAC_RX_CLK
                                                                  // 6                           GPIO6, Flash CLK
                                                                  // 7                           GPIO7, Flash D0
                                                                  // 8                           GPIO8, Flash D1
    Settings.user_template.gp.io[6]  = AGPIO(GPIO_USER);         // 9                            GPIO9, Flash D2, U1RXD
    Settings.user_template.gp.io[7]  = AGPIO(GPIO_USER);         // 10                           GPIO10, Flash D3, U1TXD
                                                                  // 11                          GPIO11, Flash CMD
    Settings.user_template.gp.io[8]  = AGPIO(GPIO_USER);         // 12      (I)O                 GPIO12, ADC2_CH5, TOUCH5, RTC_GPIO15, MTDI, HSPIQ, HS2_DATA2, SD_DATA2, EMAC_TXD3       (If driven High, flash voltage (VDD_SDIO) is 1.8V not default 3.3V. Has internal pull-down, so unconnected = Low = 3.3V. May prevent flashing and/or booting if 3.3V flash is connected and pulled high. See ESP32 datasheet for more details.)
    Settings.user_template.gp.io[9]  = AGPIO(GPIO_SWT1);         // 13   *  IO   PUMP Switch               GPIO13, ADC2_CH4, TOUCH4, RTC_GPIO14, MTCK, HSPID, HS2_DATA3, SD_DATA3, EMAC_RX_ER
    Settings.user_template.gp.io[10] = AGPIO(GPIO_PWM1);         // 14   *  IO   Servo Control             GPIO14, ADC2_CH6, TOUCH6, RTC_GPIO16, MTMS, HSPICLK, HS2_CLK, SD_CLK, EMAC_TXD2
    Settings.user_template.gp.io[11] = AGPIO(GPIO_USER);         // 15   *  (I)O HOT and Tub Water Temperature 2-DS18B20s                          GPIO15, ADC2_CH3, TOUCH3, MTDO, HSPICS0, RTC_GPIO13, HS2_CMD, SD_CMD, EMAC_RXD3         (If driven Low, silences boot messages from normal boot. Has internal pull-up, so unconnected = High = normal output.)
    Settings.user_template.gp.io[12] = AGPIO(GPIO_SWT1_NP) +2;   // 16   *  IO   Water at Drain Overflow   GPIO16, HS1_DATA4, U2RXD, EMAC_CLK_OUT
    Settings.user_template.gp.io[13] = AGPIO(GPIO_SWT1_NP) +3;   // 17   *  IO   Tub Water Overflowing     GPIO17, HS1_DATA5, U2TXD, EMAC_CLK_OUT_180
    Settings.user_template.gp.io[14] = AGPIO(GPIO_SPI_CLK);      // 18   *  IO   SPI CLK   Water in Drain  GPIO18, VSPICLK, HS1_DATA7
    Settings.user_template.gp.io[15] = AGPIO(GPIO_SPI_MISO);     // 19   *  IO   SPI MISO                  GPIO19, VSPIQ, U0CTS, EMAC_TXD0
    Settings.user_template.gp.io[16] = 0;                        // 20
    Settings.user_template.gp.io[17] = AGPIO(GPIO_DSB);          // 21   *  IO                             GPIO21, VSPIHD, EMAC_TX_EN
    Settings.user_template.gp.io[18] = AGPIO(0);                 // 22   *  IO   MCLK      LED             GPIO22, VSPIWP, U0RTS, EMAC_TXD1
    Settings.user_template.gp.io[19] = AGPIO(GPIO_SPI_MOSI);     // 23   *  IO   SPI MOSI                  GPIO23, VSPID, HS1_STROBE
    Settings.user_template.gp.io[20] = 0;                        // 24
    Settings.user_template.gp.io[21] = AGPIO(GPIO_REL1_INV) +2;  // 25   *  IO   cold water relay          GPIO25, DAC_1, ADC2_CH8, RTC_GPIO6, EMAC_RXD0
    Settings.user_template.gp.io[22] = AGPIO(GPIO_USER);         // 26   *  IO                             GPIO26, DAC_2, ADC2_CH9, RTC_GPIO7, EMAC_RXD1
    Settings.user_template.gp.io[23] = AGPIO(GPIO_USER);         // 27   *  IO                             GPIO27, ADC2_CH7, TOUCH7, RTC_GPIO17, EMAC_RX_DV
    Settings.user_template.gp.io[24] = 0;                        // 28
    Settings.user_template.gp.io[25] = 0;                        // 29
    Settings.user_template.gp.io[26] = 0;                        // 30
    Settings.user_template.gp.io[27] = 0;                        // 31
    Settings.user_template.gp.io[28] = AGPIO(GPIO_REL1_INV);     // 32   *  IO   PUMP Power relay          GPIO32, XTAL_32K_P (32.768 kHz crystal oscillator input), ADC1_CH4, TOUCH9, RTC_GPIO9
    Settings.user_template.gp.io[29] = AGPIO(GPIO_REL1_INV) +1;  // 33   *  IO   hot water relay           GPIO33, XTAL_32K_N (32.768 kHz crystal oscillator output), ADC1_CH5, TOUCH8, RTC_GPIO8
    Settings.user_template.gp.io[30] = AGPIO(GPIO_ADC_INPUT);    // 34      I   NO PULLUP  Servo Feedback  GPIO34, ADC1_CH6, RTC_GPIO4
    Settings.user_template.gp.io[31] = AGPIO(GPIO_USER);         // 35      I   NO PULLUP                  GPIO35, ADC1_CH7, RTC_GPIO5
    Settings.user_template.gp.io[32] = AGPIO(GPIO_USER);         // 36      I   NO PULLUP                  GPIO36, SENSOR_VP, ADC_H, ADC1_CH0, RTC_GPIO0
    Settings.user_template.gp.io[33] = 0;                        // 37
    Settings.user_template.gp.io[34] = 0;                        // 38
    Settings.user_template.gp.io[35] = AGPIO(GPIO_SPI_CS);       // 39      I   NO PULLUP                  GPIO39, SENSOR_VN, ADC1_CH3, ADC_H, RTC_GPIO3

But since coding to a lot deep internals this method seems to break quite often. The latest be 9.2.0.1 does not seem to like this code and is reset my configuration.

Is there a better way to build the configuration into my private builds?

thanks

**Comment by arendst** [comment, marked as answer]
Did you ever looked for command template or using define USER_TEMPLATE.

Learn how to use it once and you are done.

**Reply by mikep1998** [reply]
Is there a way to define USER_TEMPLATE using #define GPIO_USER or just the number 1? Using the numbers is not that easy.

**Comment by Jason2866** [comment]
A example is here https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L60

**Reply by mikep1998** [reply]
is there a way to write the USER_TEMPLATE using the GPIO_X enums?  Using numbers is and having to count the enums is not that easy or error free.

How about adding a #define  that would expand out in the module definitions table?  mytmplt8266 kModules8266[TMP_MAXMODULE_8285]

**Comment by Jason2866** [comment]
Use the template function and you get all values, no manual counting. Just copy and paste. See docu https://tasmota.github.io/docs/Templates/
Everything needed is already there!

**Reply by mikep1998** [reply]
Hi, thanks for the reply. While that is a process to follow its not that simple for adding to the code at compile time.

Could you provide this?


const mytmplt8266 kModules8266[TMP_MAXMODULE_8285] PROGMEM = {
  {                     // SONOFF_BASIC - Sonoff Basic (ESP8266)

**#ifdef USER_MODULE_TEMPLATE 
  USER_MODULE_TEMPLATE 
#else**

    GPI8_KEY1,          // GPIO00 Button
    GPI8_USER,          // GPIO01 Serial RXD and Optional sensor
    GPI8_USER,          // GPIO02 Only available on newer Sonoff Basic R2 V1
    GPI8_USER,          // GPIO03 Serial TXD and Optional sensor
    GPI8_USER,          // GPIO04 Optional sensor
    0,                  // GPIO05
                        // GPIO06 (SD_CLK   Flash)
                        // GPIO07 (SD_DATA0 Flash QIO/DIO/DOUT)
                        // GPIO08 (SD_DATA1 Flash QIO/DIO/DOUT)
                        // GPIO09 (SD_DATA2 Flash QIO or ESP8285)
                        // GPIO10 (SD_DATA3 Flash QIO or ESP8285)
                        // GPIO11 (SD_CMD   Flash)
    GPI8_REL1,          // GPIO12 Red Led and Relay (0 = Off, 1 = On)
    GPI8_LED1_INV,      // GPIO13 Green Led (0 = On, 1 = Off) - Link and Power status
    GPI8_USER,          // GPIO14 Optional sensor
    0,                  // GPIO15
    0,                  // GPIO16
    0                   // ADC0 Analog input

**#endif**

  },
  {                     // SONOFF_SV - Sonoff SV (ESP8266)
    GPI8_KEY1,          // GPIO00 Button
    GPI8_USER,          // GPIO01 Serial RXD and Optional sensor
    0,
    GPI8_USER,          // GPIO03 Serial TXD and Optional sensor
    GPI8_USER,          // GPIO04 Optional sensor
    GPI8_USER,          // GPIO05 Optional sensor
                        // GPIO06 (SD_CLK   Flash)
                        // GPIO07 (SD_DATA0 Flash QIO/DIO/DOUT)
                        // GPIO08 (SD_DATA1 Flash QIO/DIO/DOUT)
                        // GPIO09 (SD_DATA2 Flash QIO or ESP8285)
                        // GPIO10 (SD_DATA3 Flash QIO or ESP8285)
                        // GPIO11 (SD_CMD   Flash)
    GPI8_REL1,          // GPIO12 Red Led and Relay (0 = Off, 1 = On)
    GPI8_LED1_INV,      // GPIO13 Green Led (0 = On, 1 = Off) - Link and Power status
    GPI8_USER,          // GPIO14 Optional sensor
    0, 0,
    GPI8_USER           // ADC0 Analog input
  },


Then

#define  USER_MODULE_TEMPLATE     GPI8_KEY1,  GPI8_USER,  GPI8_USER,  GPI8_USER,  GPI8_USER,  0,  GPI8_REL1, GPI8_LED1_INV,
              GPI8_USER,  0, 0, 0 

Would produce an easy way of allowing compile time user defined modules.
Also this would take 0 byte of code.

**Reply by arendst** [reply]
All 8--bit gpi8 is deprecated and replaced by 16-bit gpio. You cannot use latest gpio addition as gpi8. gpi8 will be removed over time.

I still don't see why you cannot use already available define USER_TEMPLATE to get what you need.

**Reply by mikep1998** [reply]
I don't like define USER_TEMPLATE because coding enum numeric values is not the smartest thing to do. Prone to mistakes.  But I understand why you did it that way. I'm not complaining. I think you have done a wonderful job of coding Tasmato. 

I only gave an example of the esp8266 template. You would also need to add it to the esp32 template.

So as you more forward and change all of the 8bit code to 16bit code then the USER_TEMPLATE will look more like this:

{"NAME":"wESP32","GPIO":[1,1,1,1,1,1,0,0,0,1,1,1,5568,5600,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}

Should I be impressed with coding this way?

How about a macro that would build the template using the enums as inputs?

**Comment by Jason2866** [comment]
user_config_override.h is used at compile time and the settings here are used. This file is the place for ALL changed user settings.
I do not see any extra benefit of your suggestion. Just complicating and destroying the concept all user changes in one file.
There is ONLY Template supported for new devices. Modules are only added if code changes in Tasmota are NEEDED.
Since Template is the only supported format (for user changes) there will be no changes made for supporting old deprecated way of configuring a device.

**Reply by mikep1998** [reply]
I was not suggesting support any old formats or the use of templates. 
I was only suggesting adding this code to tasmato_template.h:

#ifdef USER_MODULE_TEMPLATE
USER_MODULE_TEMPLATE
#else

To allow for the #define USER_MODULE_TEMPLATE added to user_config_override.h file. 

Does this clear it up?
